### ISSUE TITLE
Variable type detected incorrectly inside for loop

### ISSUE THREAD

**Issue by Acidic9** [Microsoft/TypeScript#20382]
**TypeScript Version:**  2.5.2

**Code**

```ts
let myObj: { [key: number]: number } = { 0: 43, 5: 23 };

let obj = myObj[0]; // let obj: number

for (let obj in myObj) {
	if (obj == 1) {
		// ^ [ts] Operator '==' cannot be applied to types 'string' and 'number'.
		// TypeScript compiler thinks 'obj' is of type 'string' when used in `for in` loop.
	}
}
```

**Expected behavior:**

var `obj` within the `for` loop should be detected as type `number`.

**Actual behavior:**

var `obj` within the `for` loop is actually detected as type `string`.

**Comment by Acidic9** [comment]
This has now been answered here: https://stackoverflow.com/questions/47585606/typescript-compiler-not-detecting-variable-type-correctly
